[RFC] CMake: Prepare for releases.

fwalch opened the pull request:
See also #2913.

@justinmk Some comments about what should/shouldn't be in here would be nice :-) (in particular, I'm not sure if code names are a thing :-P)

Also not sure if it should work like this (i.e. build a release by specifying a certain build type), or if we want a dedicated build option (i.e. `-DRELEASE_BUILD=ON`, like `TRAVIS_BUILD`), or if we want to "hard-code" releases (e.g. remove `NVIM_VERSION_PRERELEASE` string) in a commit, tag this commit and then revert to "development mode" in a follow-up commit (i.e. set `NVIM_VERSION_PRERELEASE` back to `-alpha` and such things).

dnmario commented:
Dear Florian Walch,

The email sent to danielm@tinyhub.tk with subject
"[neovim] [WIP] CMake: Add DevRelease build type. (#3318)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

justinmk commented on CMakeLists.txt:
```diff
@@ -47,21 +47,29 @@ if(NOT CMAKE_BUILD_TYPE)
 set(NVIM_VERSION_PRERELEASE "-alpha")
-# TODO(justinmk): UTC time would be nice here #1071
-git_timestamp(GIT_TIMESTAMP)
-# TODO(justinmk): do not set this for "release" builds #1071
```
I think I have changed my mind on this, unless others object. If git _is_ available, I don't see why _not_ collect a SHA and list it in `:version`.

fwalch replied:
For almost any packaged version of Neovim, this would be "?", though, as the tarballs don't include a `.git` directory. Maybe we could change it to hide this line from `:version` if we can't detect a revision.

dnmario replied:
Dear Florian Walch,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [WIP] CMake: Prepare for releases. (#3318)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

justinmk replied:
> Maybe we could change it to hide this line from :version if we can't detect a revision.

:+1:

justinmk commented on CMakeLists.txt:
```diff
@@ -47,21 +47,29 @@ if(NOT CMAKE_BUILD_TYPE)
-set(NVIM_VERSION_MINOR 0)
+set(NVIM_VERSION_MINOR 1)
 set(NVIM_VERSION_PATCH 0)
 set(NVIM_VERSION_PRERELEASE "-alpha")
```
This will be empty string for release, of course.

Pyrohh commented on CMakeLists.txt:
```diff
@@ -47,21 +47,31 @@ if(NOT CMAKE_BUILD_TYPE)
+set(NVIM_VERSION_MINOR 1)
+set(NVIM_VERSION_PATCH 0)
+set(NVIM_VERSION_PRERELEASE "-alpha")
+set(NVIM_VERSION_CODENAME  "Whoa!")
```
Short, simple, and gets the message across; I like it.

dnmario replied:
Dear Michael Reed,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [WIP] CMake: Prepare for releases. (#3318)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

Pyrohh replied:
@dnmario I've deleted your message. I don't know if you're a bot, but please stop.

fwalch commented:
Okay.. updated this. See also neovim/homebrew-neovim#91 for the use of the build types. `RelWithDebInfo` is intended for "proper" releases (debug symbols, but no logging), whereas new `UnstableRel` is intended for developer versions (debug symbols and logging).

```
* Disable logging for release (Release, MinSizeRel, RelWithDebInfo) build
  types.
* Introduce new build type UnstableRel that replaces RelWithDebInfo for
  development builds off master and has logging enabled.

To define a release in CMake:

 1. Set NVIM_VERSION_PRERELEASE to "". Tag (annotated tag: git tag -a)
    this commit. This will:

    * Show Git tag instead of commit hash in --version. Hide commit
      information completely if we can't find any (e.g. when building
      from tarball).

 2. Set NVIM_VERSION_PRERELEASE to "-alpha". Increase version and commit.
```

fwalch commented:
`--version` output for a prerelease:

```
NVIM 0.1.0-alpha+201509271233 (compiled Sep 27 2015 12:44:01)
Commit: 706201da11e6378141c114b01e8f6e6f59f9f922
Build type: UnstableRel
Compilation: ....
```

`--version` output for a release, when currently checked out revision has been tagged with `git tag -a v0.1`:

```
NVIM 0.1.0 (compiled Sep 27 2015 16:43:58)
Commit: v0.1
Build type: RelWithDebInfo
Compilation: ...
```

`--version` output for a release, when Git info can't be found (tarball) or currently checked out revision does not have a tag:
```
NVIM 0.1.0 (compiled Sep 27 2015 16:42:03)
Build type: RelWithDebInfo
Compilation: ...
```

justinmk commented:
What happened to `DevRelease`?

fwalch commented:
> What happened to DevRelease?

Oh :-D I just thought that this was too confusing a name.. in the first iteration of the PR, I actually confused the flags of RelWithDebInfo (no logging, debug info) and DevRelease (logging, debug info) myself. But I can change it back, of course.

justinmk commented:
How about `RelDev`

fwalch commented:
@justinmk Updated.

justinmk commented on CMakeLists.txt:
```diff
@@ -38,31 +38,48 @@ endif()
+#  2. Set NVIM_VERSION_PRERELEASE to "-alpha". Increase version and commit.
 set(NVIM_VERSION_MAJOR 0)
-set(NVIM_VERSION_MINOR 0)
+set(NVIM_VERSION_MINOR 1)
```
If you leave this at 0 we can merge this now...

fwalch replied:
Fixed.

fwalch commented:
Moved the `file(TO_NATIVE_PATH ${CMAKE_CURRENT_LIST_DIR}/.git GIT_DIR)` out of the `if()`, I think this should work with `git_get_exact_tag` as well.

justinmk commented on src/nvim/version.c:
```diff
@@ -36,7 +36,9 @@ char *Version = VIM_VERSION_SHORT;
 char *longVersion = NVIM_VERSION_LONG;
 char *longVersionWithDate = NVIM_VERSION_LONG " (compiled " __DATE__ " " __TIME__ ")";
 char *mediumVersion = NVIM_VERSION_MEDIUM;
+#ifdef NVIM_VERSION_HAS_COMMIT
```
Can this just check `#ifdef NVIM_VERSION_COMMIT` ?

fwalch replied:
Don't think so; see https://github.com/neovim/neovim/pull/3318#discussion_r40505670.

fwalch replied:
Okay, changed this around and put an `ifdef VERSION_HAS_COMMIT` in `versiondef.h.in`.

justinmk commented on CMakeLists.txt:
```diff
@@ -38,31 +38,48 @@ endif()
+if(NVIM_VERSION_PRERELEASE)
+  get_git_head_revision(GIT_REFSPEC NVIM_VERSION_COMMIT)
+  if(NOT NVIM_VERSION_COMMIT)
+    set(NVIM_VERSION_COMMIT "?")
```
I think we decided to leave this out. If commit not found, that line shouldn't appear in `:version`.

fwalch replied:
I thought that was only for "releases", but I'll change it.

justinmk commented on CMakeLists.txt:
```diff
@@ -38,31 +38,48 @@ endif()
+  if(NOT NVIM_VERSION_COMMIT)
+    set(NVIM_VERSION_COMMIT "?")
+  endif()
+  set(NVIM_VERSION_HAS_COMMIT 1)
```
Trying to cut down on the cmake code, can we avoid this extra variable?

fwalch replied:
I don't think so.. in `versiondef.h.in`, we have:

```
#define NVIM_VERSION_COMMIT "@NVIM_VERSION_COMMIT@"
´``

So `NVIM_VERSION_COMMIT` (in C) will be an empty string if we don't have found any commit in CMake. So it's always defined, and checking for an empty string in C macros is apparently not that easy.

Maybe there's something like `#cmakeif`, like `#cmakedefine`...

justinmk commented on CMakeLists.txt:
```diff
@@ -71,6 +88,54 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+      CACHE STRING "Flags used for linking binaries during development (optimized, but with debug info and logging) builds."
+      FORCE)
+  SET(CMAKE_SHARED_LINKER_FLAGS_DEV ""
+      CACHE STRING "Flags used by the shared libraries linker during development (optimized, but with debug info and logging) builds."
```
Can this code that is duplicated inside the `if(MSVC)` block be done outside the condition to avoid the duplicate code?

fwalch replied:
Done.

fwalch commented:
Updated (also see updated commit message).

fwalch commented:
@jszakmeister We have the following in `versiondef.h.in`:

```c
#define NVIM_VERSION_COMMIT "@NVIM_VERSION_COMMIT@"
#cmakedefine NVIM_VERSION_HAS_COMMIT
```

and then use this similar to this in `version.c`:

```c
#ifdef NVIM_VERSION_HAS_COMMIT
printf("Commit: " NVIM_VERSION_COMMIT);
#endif
```

We want to hide the `Commit: ...` line completely if `NVIM_VERSION_COMMIT` is an empty string. We have something in CMake that does this by setting `NVIM_HAS_VERSION`:

```cmake
if(NVIM_VERSION_COMMIT)
  set(NVIM_VERSION_HAS_COMMIT 1)
else()
  set(NVIM_VERSION_HAS_COMMIT 0)
endif()
```

Do you know of an easier way to solve this, without having two defines `NVIM_VERSION_COMMIT` and `NVIM_VERSION_HAS_COMMIT`?

justinmk commented:
I was just reminded of [git describe](https://git-scm.com/book/en/v2/Distributed-Git-Maintaining-a-Project#Generating-a-Build-Number). After we get our first tag, we should use that for the `Dev` version number instead of hard-coding `x.x-alpha`.

fwalch commented:
@justinmk huge :+1:
I'll be using that for the Arch PKGBUILD and was already worried that there would be a mismatch between --version and the Arch package.

fwalch committed:
CMake: Add custom Dev build type.

Introduce new build type Dev that replaces RelWithDebInfo for development
builds off master and has optimizations, debug info, and logging enabled.
Keep assertions enabled for RelWithDebInfo.

fwalch committed:
CMake: Disable logging for release build types.

These are: Release, MinSizeRel, and RelWithDebInfo.

Closes #2913.

fwalch committed:
version: Prepare for releases.

* Hide commit information from --version if we can't find any (e.g. when
  building from tarball).

To define a release in CMake, set NVIM_VERSION_PRERELEASE to "".
This will modify --version output to:

    * Show annotated Git tag instead of commit hash (NVIM_VERSION_COMMIT).
    * Hide commit date (NVIM_VERSION_BUILD).

fwalch commented:
Split into multiple commits and re-enabled assertions for RelWithDebInfo by removing `-DNDEBUG` from its default flags.

fwalch commented:
@justinmk Fixed the "two #define problem".. actually surprisingly easy if you read the docu carefully :grin:

justinmk commented:
Yay! :)

fwalch committed:
CMake: Force use of project directory to look for Git data.

Before this change, building Neovim would recursively search parent
directories for a .git directory. If Neovim was downloaded as a tarball
(i.e. without a .git directory), but placed in a subdirectory of
a Git repository, this caused a CMake error. Such a situation could
occur when packaging Neovim, for example.

Unfortunately, the previous attempt in #3317 did not fix this problem.

fwalch commented:
Testing the packaging process for Arch, I found out that #3317 did, in fact, not do anything. Pushed a commit that should fix the underlying problem for real this time.

fwalch commented:
Found out that having *both* `NDEBUG` and `DISABLE_LOG` defined didn't play well together.. fixed.

Sorry, I know that for an [RFC] PR, there's still an awful lot of changes going on.. but I think this might be it now.

fwalch commented:
Any more comments? :-)

justinmk commented:
I will take a close look later today :)

fwalch commented:
Rebased.

fwalch commented:
Rebased.

jrobeson commented on CMakeLists.txt:
```diff
@@ -38,31 +38,36 @@ endif()
 if(NOT CMAKE_BUILD_TYPE)
-  message(STATUS "CMAKE_BUILD_TYPE not given; setting to 'RelWithDebInfo'.")
-  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Choose the type of build." FORCE)
+  message(STATUS "CMAKE_BUILD_TYPE not given; setting to 'Dev'.")
```
change the worlding to "defaulting to 'Dev'" ?

fwalch replied:
:+1:

tarruda commented:
LGTM. After this is merged I think we can release 0.1

fwalch commented:
Rebased and fixed a typo (`FORCE_GIT_DIR` vs `FORCED_GIT_DIR`).

justinmk commented:
Testing this now.

justinmk commented:
Thank you @fwalch
